ice-adapter/src/main/java/com/faforever/iceadapter/ice/GameSession.java

```diff
@@ -125,30 +126,50 @@ public CompletableFuture<OptionalDouble> load(String host) throws Exception {
                     urls = Collections.singletonList((String) iceServerData.get("url"));
                 }
 
-                urls.stream()
-                        .map(IceServer.urlPattern::matcher)
-                        .filter(Matcher::matches)
-                        .forEach(matcher -> {
-                            String host = matcher.group("host");
-                            int port = Optional.ofNullable(matcher.group("port")).map(Integer::parseInt).orElse(3478);
-                            Transport transport = Optional.ofNullable(matcher.group("transport")).map(Transport::parse).orElse(Transport.UDP);
-
-                            TransportAddress address = new TransportAddress(host, port, transport);
-                            (matcher.group("protocol").equals("stun") ? iceServer.getStunAddresses() : iceServer.getTurnAddresses()).add(address);
+                urls.stream().map(stringUrl -> {
+                    try {
+                        return new URL(stringUrl);
+                    } catch (Exception e) {
+                        log.warn("Invalid ICE server URL: {}", stringUrl);
+                        return null;
+                    }
+                }).filter(Objects::nonNull).forEach(url -> {
+                    String host = url.getHost();
+                    int port = url.getPort() == -1 ? 3478 : url.getPort();
+                    Transport transport = Optional.ofNullable(url.getQuery())
+                                                  .stream()
+                                                  .flatMap(query -> Arrays.stream(query.split("&")))
+                                                  .map(param -> param.split("="))
+                                                  .filter(param -> param.length == 2)
+                                                  .filter(param -> param[0].equals("transport"))
+                                                  .map(param -> param[1])
+                                                  .map(Transport::parse)
+                                                  .findFirst()
```

Comment:
This can be moved up 2 lines.

Reply:
Being up or down 2 lines doesn't matter the find first is an early termination operation so won't run anymore operations once a matching one is found.

I originally had it two lines up but decided I liked this better.